README.md
# synadart

The `synadart` library can be used to create neural networks of any complexity,
as well as learn from the source code by studying its extremely clean
implementation.

## Launching our first network

To begin using the `synadart`, we must first import it into our project:

```dart
import 'package:synadart/synadart.dart';
```

Next, we must create a network of our chosen type. Let's create a sequential
network, in which every layer has one input and one output tensor. This should
be pretty easy:

```dart
final network = Sequential(learningRate: 0.3);
```

Our network is currently empty; it contains no layers and therefore no neurons.
Let's add three layers; the input layer, one hidden layer and the output layer:

```dart
network.addLayers([
  Dense(15, activationAlgorithm: ActivationAlgorithm.sigmoid),
  Dense(5, activationAlgorithm: ActivationAlgorithm.sigmoid),
  Dense(1, activationAlgorithm: ActivationAlgorithm.sigmoid),
]);
```

Now that our network has some structure to it, we can begin using it.. No, not
quite yet. Our network is still not trained, and has no clue what it is doing.
Time to train it.

Firstly, we will create a list of _expected_ values, i.e. _values we are
expecting the network to output_. Here, we are expecting to get the number '5'.

```dart
final expected = [
  [0.01], // 0
  [0.01], // 1
  [0.01], // 2
  [0.01], // 3
  [0.01], // 4
  [0.99], // 5 ( This is what we are anticipating )
  [0.01], // 6
  [0.01], // 7
  [0.01], // 8
  [0.01], // 9
];
```

Fantastic, we are now expecting our infantile network to magically output a
number 5, not having taught it a thing. Oh, right - that's where the training
data part comes in!

We must now tell the network what each of our expected output values is
associated with. Let's teach it some numbers:

```dart
final trainingData = [
  '111101101101111'.split('').map(double.parse).toList(), // Pixel representation of a 0,
  '001001001001001'.split('').map(double.parse).toList(), // a 1,
  '111001111100111'.split('').map(double.parse).toList(), // a 2,
  '111001111001111'.split('').map(double.parse).toList(), // a 3,
  '101101111001001'.split('').map(double.parse).toList(), // a 4,
  '111100111001111'.split('').map(double.parse).toList(), // a 5,
  '111100111101111'.split('').map(double.parse).toList(), // a 6,
  '111001001001001'.split('').map(double.parse).toList(), // a 7,
  '111101111101111'.split('').map(double.parse).toList(), // an 8,
  '111101111001111'.split('').map(double.parse).toList(), // and a 9.
];
```

Now that we granted our network a grand total of 10 numbers to learn, we can
begin training the network using the values we've set up:

```dart
network.train(inputs: trainingData, expected: expected, iterations: 5000);
```

Wonderful! We've trained our network using the pixel representation of number
images, and our network is now able to recognise the number '5' with relative
confidence. The last step is to test our network's capabilities ourselves.

Let's give our network a couple pixel representations of distorted images of the
number '5':

```dart
final testData = [
  '111100111000111'.split('').map(double.parse).toList(),
  '111100010001111'.split('').map(double.parse).toList(),
  '111100011001111'.split('').map(double.parse).toList(),
  '110100111001111'.split('').map(double.parse).toList(),
  '110100111001011'.split('').map(double.parse).toList(),
  '111100101001111'.split('').map(double.parse).toList(),
];
```

To check the confidence of the network in recognising distorted '5's:

```dart
for (final test in testData) {
  print('Confidence in recognising a distorted 5: ${network.process(test)}');
}
```


CHANGELOG.md
## 0.5.0

- Update licence bearer and links.
- Bump dependencies.
- Bump the SDK version to `3.0.0`.
- BREAKING: Remove the `sprint` dependency, which is inappropriate for a package of this
  kind. Replaced logging with documented exceptions.

## 0.4.5

- Relicensed from GPLv3 to MIT.

## 0.4.4

- Update SDK version from `2.12.0` to `2.17.0`.
- Updated project description to make it more accurate in describing what
  `synadart` actually is.

## 0.4.3

- Bumped version of `sprint` from `1.0.2+3` to `1.0.3`.
- Updated repository, homepage and issue tracker links.
- Refactored and made formatting and style changes to bring the project up to
  par.

## 0.4.2+1

- Updated package description.

## 0.4.2

- Updated `sprint` version from `1.0.0+1` to `1.0.2+3`.
- Replaced the now discontinued `pedantic` with the `words` lint ruleset.
- Reversed the order of versions in `CHANGELOG.md` from ascending to descending.

## 0.4.1+1

- Refactored code.
- Removed `logger.dart` in favour of the `sprint` package.

## 0.4.1

- Updated documentation.

## 0.4.0

- Organised code.
- Replaced network types such as `feed-forward` or `deep feed-forward` with a
  single class `Sequential`.
- Moved focus from `Network` to `Layer`, so that different layers can be added
  to a `Network`, rather than creating new types of networks, and limiting the
  user to a preset model.
- Updated `example.dart` and `README.md`.

## 0.3.2

- Added a simple feed-forward network model.

## 0.3.1

- Added 5 new activation functions: `SeLU`, `Softplus`, `Softsign`, `Swish` and
  `Gaussian`.
- Renamed the 'logistic' function to 'sigmoid'.
- Created function `abs()` for obtaining the absolute value of a variable.

## 0.3.0

- Updated documentation of `Logger`, `Backpropagation` and `ValueGenerator`.
- Created `/examples` directory with a file `example.dart` that displays the
  network being used to recognise the number '5'.

## 0.2.5

- Renamed 'Multilayer Perceptron' to 'Deep Feed-Forward', since 'deep
  feed-forward' is broader as a concept than 'multi-layer perceptrons'.

## 0.2.4

- Updated documentation of `activation.dart`, having added explanations for the
  different activation functions.

## 0.2.3

- Updated documentation of `Network`.
- Replaced `process()` in `Layer` with an `output` getter, simplifying the
  method of getting each `Neuron`'s output.

## 0.2.2

- Updated documentation of `Layer` and removed a chunk of dead code.

## 0.2.1

- Removed the feed-forward network and simple perceptrons in favour of an
  upcoming simpler implementation of networks, through the use of a single
  network model.
- Added [learningRate] as a parameter, and removed the hard-coded value of
  `0.2`.
- Updated documentation of `Neuron`.

## 0.2.0

- Added a feed-forward network and simple perceptrons.
- Added `LReLU`, `eLU` and `tanh` activation functions.
- Renamed 'sigmoid' to 'logistic'.

## 0.1.1

- Added `README.md` and updated formatting.

## 0.1.0

- Implemented a multilayer perceptron and a basic algorithm for backpropagation.


example/recognise_5.dart
import 'package:synadart/src/layers/dense.dart';
import 'package:synadart/synadart.dart';

void main() {
  final network = Sequential(learningRate: 0.3);
  
  network.addLayer(Dense(
    size: 15,
    activationAlgorithm: ActivationAlgorithm.sigmoid,
  ));

  network.addLayer(Dense(
    size: 5,
    activationAlgorithm: ActivationAlgorithm.sigmoid,
  ));

  network.addLayer(Dense(
    size: 1,
    activationAlgorithm: ActivationAlgorithm.sigmoid,
  ));

// We are expecting to get the number '5'
final expected = [
  [0.01],
  [0.01],
  [0.01],
  [0.01],
  [0.01],
  [0.99],
  [0.01],
  [0.01],
  [0.01],
  [0.01],
];

  // Training data contains different number patterns
final trainingData = [
  '111101101101111'.split('').map(double.parse).toList(),
  '001001001001001'.split('').map(double.parse).toList(),
  '111001111100111'.split('').map(double.parse).toList(),
  '111001111001111'.split('').map(double.parse).toList(),
  '101101111001001'.split('').map(double.parse).toList(),
  '111100111001111'.split('').map(double.parse).toList(), // This is the number 5
  '111100111101111'.split('').map(double.parse).toList(),
  '111001001001001'.split('').map(double.parse).toList(),
  '111101111101111'.split('').map(double.parse).toList(),
  '111101111001111'.split('').map(double.parse).toList(),
];

  // Test data which contains distorted patterns of the number 5
final testData = [
  '111100111000111'.split('').map(double.parse).toList(),
  '111100010001111'.split('').map(double.parse).toList(),
  '111100011001111'.split('').map(double.parse).toList(),
  '110100111001111'.split('').map(double.parse).toList(),
  '110100111001011'.split('').map(double.parse).toList(),
  '111100101001111'.split('').map(double.parse).toList(),
];

  // The number 5 itself
  final numberFive = trainingData[5];

  // Train the network using the training and expected data
  network.train(inputs: trainingData, expected: expected, iterations: 5000);

  print('Confidence in recognising a 5: ${network.process(numberFive)}');
for (final test in testData) {
  print('Confidence in recognising a distorted 5: ${network.process(test)}');
}
  print('Is 0 a 5? ${network.process(trainingData[0])}');
  print('Is 8 a 5? ${network.process(trainingData[8])}');
  print('Is 3 a 5? ${network.process(trainingData[3])}');
}
